staging/src/k8s.io/endpointslice/topologyheuristics/manager.go

```diff
@@ -0,0 +1,344 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package topologyheuristics
+
+import (
+	"fmt"
+	"slices"
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+	discoveryv1 "k8s.io/api/discovery/v1"
+	"k8s.io/klog/v2"
+)
+
+// Manager manages multiple EndpointSlice Topology Heuristics. It exposes
+// functions for adding and removing topology hints to endpoint slices using
+// various heuristics.
+type Manager struct {
+	// heuristicsByName maintains a mapping of a heuristic name to a Heuristic
+	// instance.
+	heuristicsByName map[string]Heuristic
+	// defaultHeuristicName is the name of the Heuristic which will be considered
+	// the default.
+	defaultHeuristicName string
+	// nonEPSHintHeuristicNames should contain the names of heuristics which are not
+	// implemented through EndpointSlice hints and may be implemented in the
+	// Dataplane (or in other ways). Manager will NOT send "unrecognizable" events
+	// for these heuristic names (despite the fact that it does not populate
+	// EndpointSlice hints for these.)
+	nonEPSHintHeuristicNames []string
+
+	mu sync.Mutex
+	// activeHeuristicsByService keeps track of the heuristic currently in use by
+	// a Service.
+	activeHeuristicsByService map[string]map[discoveryv1.AddressType]Heuristic
+}
+
+// NewManager returns an instance of a heuristic Manager.
+//
+// `heuristics` contains the set of heuristics that this Mangaer will use and
+// manage. They should have unique names (as returned by their Heuristic.Name()
```

Comment:
Are these names things that implementations can define?  How do we "reserve" names so they don't collide?